controllers/nodefeaturediscovery_status.go

```diff
@@ -0,0 +1,525 @@
+			LastTransitionTime: metav1.Time{Time: now},
+		},
+		{
+			Type:               ConditionUpgradeable,
+			Status:             metav1.ConditionFalse,
+			LastTransitionTime: metav1.Time{Time: now},
+		},
+		{
+			Type:               ConditionProgressing,
+			Status:             metav1.ConditionFalse,
+			LastTransitionTime: metav1.Time{Time: now},
+		},
+		{
+			Type:               ConditionDegraded,
+			Status:             metav1.ConditionTrue,
+			LastTransitionTime: metav1.Time{Time: now},
+			Reason:             reason,
+			Message:            message,
+		},
+	}
+}
+
+// getProgressingConditions returns a list of conditions.Condition objects and marks
+// every condition as FALSE except for conditions.ConditionProgressing so that the
+// reconciler can determine that the resource is progressing.
+func (r *NodeFeatureDiscoveryReconciler) getProgressingConditions(reason string, message string) []metav1.Condition {
+	now := time.Now()
+	return []metav1.Condition{
+		{
+			Type:               ConditionAvailable,
+			Status:             metav1.ConditionFalse,
+			LastTransitionTime: metav1.Time{Time: now},
+		},
+		{
+			Type:               ConditionUpgradeable,
+			Status:             metav1.ConditionFalse,
+			LastTransitionTime: metav1.Time{Time: now},
+		},
+		{
+			Type:               ConditionProgressing,
+			Status:             metav1.ConditionTrue,
+			LastTransitionTime: metav1.Time{Time: now},
+			Reason:             reason,
+			Message:            message,
+		},
+		{
+			Type:               ConditionDegraded,
+			Status:             metav1.ConditionFalse,
+			LastTransitionTime: metav1.Time{Time: now},
+		},
+	}
+}
+
+// The status of the resource (available, upgradeable, progressing, or
+// degraded).
+type Status struct {
+
+	// Is the resource available, upgradable, etc.?
+	isAvailable   bool
+	isUpgradeable bool
+	isProgressing bool
+	isDegraded    bool
+}
+
+// initializeDegradedStatus initializes the status struct to degraded
+func initializeDegradedStatus() Status {
+	return Status{
+		isAvailable:   false,
+		isUpgradeable: false,
+		isProgressing: false,
+		isDegraded:    true,
+	}
+}
+
+// getWorkerDaemonSetConditions is a wrapper around "getDaemonSetConditions" for
+// worker DaemonSets
+func (r *NodeFeatureDiscoveryReconciler) getWorkerDaemonSetConditions(ctx context.Context) (Status, error) {
+	return r.getDaemonSetConditions(ctx, workerInstance)
+}
+
+// getMasterDaemonSetConditions is a wrapper around "getDaemonSetConditions" for
+// master DaemonSets
+func (r *NodeFeatureDiscoveryReconciler) getMasterDaemonSetConditions(ctx context.Context) (Status, error) {
+	return r.getDaemonSetConditions(ctx, masterInstance)
+}
+
+// getDaemonSetConditions gets the current status of a DaemonSet. If an error
+// occurs, this function returns the corresponding error message
+func (r *NodeFeatureDiscoveryReconciler) getDaemonSetConditions(ctx context.Context, nodeType string) (Status, error) {
```

Comment:
```suggestion
func (r *NodeFeatureDiscoveryReconciler) getDaemonSetConditions(ctx context.Context, appName string) (Status, error) {
```

and ditto X times below...

You can come up with a better name than ` appName`, but something other than `nodeType`please 😄

Reply:
how about `nfdRole` master and worker is a role they play on the deployment, no?

Reply:
This reminds too much RBAC IMO, roles and rolebindings...

Dunno, what would be the best. appName, instanceType, daemonType, instanceName... 🙃